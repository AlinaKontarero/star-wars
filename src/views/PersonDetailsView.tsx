import * as React from 'react'
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome"
import { Tooltip, IconButton } from '@material-ui/core'
import CloseIcon from '@material-ui/icons/Close';
import Chip from '../components/Chip'
import { Zoom } from '@material-ui/core';
import Loading from '../components/Loading';
import { IPerson } from '../redux/reducers/personReducer';
import { IconName } from '@fortawesome/free-solid-svg-icons';
import { IFilm } from './View';

interface Props {
  person: IPerson
  onClose: () => void 
  films: IFilm[]
}

class PersonDetailsView extends React.Component<Props, never> { 
  render() {
    const { person, films, onClose } = this.props
    const genderIcon = (): IconName => {
      switch(person.gender) {
        case 'male':
          return 'mars'
        case 'female':
          return 'venus'
        case 'n/a':
        default:
          return 'genderless'
      }
    }

    const setFilmTitles = (url: string) =>  films.find(film => film.url === url)

    const filmsSection = (): JSX.Element => {
      const filmsArr: IFilm[] = []
      const filteredByUrlFilms = person.films.map(_movie => filmsArr.push(setFilmTitles(_movie) || {} as IFilm))
      return filmsArr.length === 0 
        ? <Loading />
        : <> {filmsArr.map((_film: IFilm, index: number) => (
           <div className='column is-narrow' key={index}>
            <Chip label={_film.title}/>
          </div>
          ))}
          </>
    }

    const rowWrapper = (key: string, value: string | number | JSX.Element) => {
      return (
        <div className='column is-full'> 
          <p><b>{key}: </b><span>{value}</span></p>
        </div>
      )
    }

    return (
      <div className='column is-full'>
        <div className='columns is-multiline is-variable is-2 is-vcentered'>
          <div className='column'>
            <h2>Person details</h2>
          </div>
          <div className='column is-narrow'>
          <Tooltip title='Close' TransitionComponent={Zoom} arrow={true} placement={'top'}>
            <IconButton 
              onClick={onClose} 
              color='inherit'
              >
              <CloseIcon />
            </IconButton>
          </Tooltip>
          </div>
          {rowWrapper('Name', person.name)}
          {rowWrapper('Birth year', person.birth_year)}
          {rowWrapper('Gender', <FontAwesomeIcon icon={["fas", genderIcon() ]} size='lg' />)}
          {filmsSection()}
        </div>
      </div>
    )
  }
}

export default PersonDetailsView