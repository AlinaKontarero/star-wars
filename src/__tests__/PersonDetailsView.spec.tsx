import React from 'react';
import { Provider } from 'react-redux';
import renderer, { ReactTestRenderer } from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import PersonDetailsView from '../views/PersonDetailsView';
import { mockFilms } from './mockups/mockFilms';
import { mockPersons } from './mockups/mockPersons';
 
const mockStore = configureStore([]);
 
describe('PersonDetailsView Component', () => {
  let store;
  let component: ReactTestRenderer
 
  beforeEach(() => {
    store = mockStore({
      PersonState: {
        Persons: mockPersons,
      } 
    });
 
    component = renderer.create(
      <Provider store={store}>
        <PersonDetailsView films={mockFilms} person={mockPersons[0]} onClose={() => console.log('close')} />
      </Provider>
    );
  });
  
  
  it('should render with given state from Redux store', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });
});