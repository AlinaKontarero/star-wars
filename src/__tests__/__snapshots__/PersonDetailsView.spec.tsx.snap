// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PersonDetailsView Component should render with given state from Redux store 1`] = `
<div
  className="column is-full person-details-section"
>
  <div
    className="columns is-multiline is-variable is-2 is-vcentered"
  >
    <div
      className="column"
    >
      <h2>
        Person details
      </h2>
    </div>
    <div
      className="column is-narrow"
    >
      <button
        aria-describedby={null}
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        title="Close"
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      className="column is-full"
    >
      <p>
        <b>
          Name
          : 
        </b>
        <span>
          Luke Skywalker
        </span>
      </p>
    </div>
    <div
      className="column is-full"
    >
      <p>
        <b>
          Birth year
          : 
        </b>
        <span>
          19BBY
        </span>
      </p>
    </div>
    <div
      className="column is-full"
    >
      <p>
        <b>
          Gender
          : 
        </b>
        <span />
      </p>
    </div>
     
    <div
      className="column is-narrow"
    >
      <div
        className="MuiChip-root makeStyles-root-1 MuiChip-outlined"
        onKeyDown={[Function]}
        onKeyUp={[Function]}
      >
        <i
          className="MuiChip-icon makeStyles-icon-2 fas fa-film fa-lg"
        />
        <span
          className="MuiChip-label"
        >
          A New Hope
        </span>
      </div>
    </div>
    <div
      className="column is-narrow"
    >
      <div
        className="MuiChip-root makeStyles-root-1 MuiChip-outlined"
        onKeyDown={[Function]}
        onKeyUp={[Function]}
      >
        <i
          className="MuiChip-icon makeStyles-icon-2 fas fa-film fa-lg"
        />
        <span
          className="MuiChip-label"
        >
          The Empire Strikes Back
        </span>
      </div>
    </div>
    <div
      className="column is-narrow"
    >
      <div
        className="MuiChip-root makeStyles-root-1 MuiChip-outlined"
        onKeyDown={[Function]}
        onKeyUp={[Function]}
      >
        <i
          className="MuiChip-icon makeStyles-icon-2 fas fa-film fa-lg"
        />
        <span
          className="MuiChip-label"
        />
      </div>
    </div>
    <div
      className="column is-narrow"
    >
      <div
        className="MuiChip-root makeStyles-root-1 MuiChip-outlined"
        onKeyDown={[Function]}
        onKeyUp={[Function]}
      >
        <i
          className="MuiChip-icon makeStyles-icon-2 fas fa-film fa-lg"
        />
        <span
          className="MuiChip-label"
        >
          Revenge of the Sith
        </span>
      </div>
    </div>
  </div>
</div>
`;
